File: DSA Implementations/Python/quick_sort.py
def pivot_index(arr, low, high):
    
    pivot = arr[low]
    i = low
    j= high
    
    while i < j:
        
        while i <= high and arr[i] <= pivot:
            i += 1
        
        while j >= low and arr[j] > pivot:
            j -= 1
        
        if i < j:
            arr[i], arr[j] = arr[j], arr[i]
    
    arr[low], arr[j] = arr[j], arr[low]
    
    return j

def quick_sort(arr, low, high):
    if low < high:
        pIndex = pivot_index(arr, low, high)
        quick_sort(arr, low, pIndex-1)
        quick_sort(arr, pIndex+1, high)

n = int(input("Enter the number of elements in the array : "))
arr = []

low = 0
high = n-1

print("Enter the elements of the array : ")
for _ in range(n):
    arr.append(int(input()))

quick_sort(arr, low, high)

print("Sorted array : ")
print(*arr)

File: DSA Implementations/C++/quick_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int partition(vector<int> &arr, int low, int high)
{
    int pivot = arr[low];
    int i = low;
    int j = high;

    while (i < j)
    {
        while (arr[i] <= pivot && i <= high - 1)
        {
            i++;
        }
        while (arr[j] > pivot && j >= low)
        {
            j--;
        }
        if (i < j)
        {
            swap(arr[i], arr[j]);
        }
    }
    swap(arr[low], arr[j]);
    return j;
}

void quick_sort(vector<int> &arr, int low, int high)
{
    if (low < high)
    {
        int p_index = partition(arr, low, high);
        quick_sort(arr, low, p_index - 1);
        quick_sort(arr, p_index + 1, high);
    }
}

int main()
{
    int n;
    cout << "Enter the number of elements in the array: ";
    cin >> n;

    vector<int> arr(n);
    cout << "Enter the elements of the array: " << endl;
    for (int i = 0; i < n; ++i)
    {
        cin >> arr[i];
    }

    quick_sort(arr, 0, n - 1);

    cout << "Sorted array: " << endl;
    for (int i = 0; i < n; ++i)
    {
        cout << arr[i] << " ";
    }
    cout << endl;

    return 0;
}